File: backend/task-service/src/main/java/com/example/task_service/services/TaskService.java
package com.example.task_service.services;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import com.example.task_service.repository.TaskRepository;
import com.example.task_service.entities.Task;

import java.time.LocalDateTime;
import java.util.List;

/**
 * Service de gestion des tâches.
 * 
 * Ce service fournit des méthodes pour effectuer des opérations sur les tâches, comme la création, 
 * la récupération, la mise à jour et la suppression de tâches. Il interagit avec le {@link TaskRepository} 
 * pour accéder à la base de données.
 * 
 * @author Courné Jules
 * @version 1.0
 */
@Service
public class TaskService {

    @Autowired
    private TaskRepository taskRepository;

    /**
     * Récupère toutes les tâches disponibles dans la base de données.
     * 
     * @return Une liste contenant toutes les tâches.
     */
    public List<Task> getAllTasks() {
        return taskRepository.findAll();
    }

    /**
     * Crée une nouvelle tâche dans la base de données.
     * 
     * @param task L'objet Task à créer.
     * @return La tâche qui a été enregistrée dans la base de données.
     */
    public Task createTask(Task task) {
        return taskRepository.save(task);
    }

    /**
     * Récupère une tâche par son identifiant.
     * 
     * @param id L'identifiant de la tâche à récupérer.
     * @return La tâche correspondante.
     * @throws RuntimeException Si aucune tâche n'est trouvée avec l'ID donné.
     */
    public Task getTaskById(Long id) {
        return taskRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Task not found"));
    }

    /**
     * Récupère toutes les tâches associées à un utilisateur spécifique via son email.
     * 
     * @param userEmail L'email de l'utilisateur pour lequel récupérer les tâches.
     * @return Une liste de tâches associées à cet utilisateur.
     * @throws RuntimeException Si aucune tâche n'est trouvée pour cet utilisateur.
     */
    public List<Task> getTaskByUserEmail(String userEmail) {
        List<Task> tasks = taskRepository.findByUserMail(userEmail); // Mise à jour pour utiliser userMail
        if (tasks.isEmpty()) {
            throw new RuntimeException("No tasks found for user with email: " + userEmail);
        }
        return tasks;
    }

    /**
     * Supprime une tâche en fonction de son identifiant.
     * 
     * @param id L'identifiant de la tâche à supprimer.
     */
    public void deleteTask(Long id) {
        taskRepository.deleteById(id);
    }

    /**
     * Récupère une tâche en fonction de sa date de création.
     * 
     * @param createDate La date de création de la tâche à rechercher.
     * @return La tâche correspondante à la date de création donnée.
     */
    public Task getTaskByCreateDate(LocalDateTime createDate) {
        return taskRepository.findByCreateDate(createDate);
    }

    /**
     * Met à jour une tâche existante.
     * 
     * @param task L'objet Task contenant les nouvelles informations pour la tâche à mettre à jour.
     * @return La tâche mise à jour.
     */
    public Task updateTask(Task task) {
        return taskRepository.save(task);
    }

    /**
     * Récupère les tâches associées à un projet donné par son identifiant.
     * 
     * @param projectId L'identifiant du projet pour lequel récupérer les tâches.
     * @return Une liste de tâches associées au projet.
     */
    public List<Task> getTaskByProjectId(Long projectId) {
        return taskRepository.findByProjectId(projectId);
    }
}

File: frontend/src/service/TaskService.js
import axios from 'axios';

// Constantes de configuration de l'url pour accéder au service de gestion des tâches
const url = "http://localhost:8080/api/tasks";

/**
 * Fonction qui retourne le token d'authentification
 * 
 * @returns le token jwt de l'utilisateur
 */
const getAuthToken = () => {
  return sessionStorage.getItem('jwtToken');
};

/**
 * Récupère la liste des tâches
 * 
 * @returns response.data - list of tasks
 */
export async function getAllTasks() {
  try {
    const response = await axios.get(url, {
      headers: {
        'Authorization': `Bearer ${getAuthToken()}`
      }
    });
    return response.data; 
  } catch (error) {
    console.error('Erreur lors de la récupération des tâches:', error.response ? error.response.data : error.message);
    throw error; 
  }
}

/**
 * Créer une nouvelle tâche
 * 
 * @param {Object} task - la tâche à créer
 * @returns response.data - la tâche crée
 */
export async function createTask(task) {
  try {
    const response = await axios.post(`${url}/create`, task, {
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${getAuthToken()}`
      }
    });

    console.log('Tâche créée avec succès:', response.data);
    return response.data; 
  } catch (error) {
    console.error('Erreur de création de tâche:', error.response ? error.response.data : error.message);
    throw error; 
  }
}

/**
 * Modifie une tâche
 * 
 * @param {Date} createDate - la date de création de la tâche à modifier
 * @param {Object} updatedTicket - la tâche modifié
 * @returns la tâche modifié
 */
export const updateTask = async (createDate, updatedTicket) => {

  return await axios.put(`${url}/update?createDate=${createDate}`, updatedTicket, {
      headers: {
          'Authorization': `Bearer ${getAuthToken()}`,
          'Content-Type': 'application/json',
      },
  });
};

/**
 * Récupère la tâche avec l'id id
 * 
 * @param {number} id - l'id de la tâche à récupérer
 * @returns response.data - la tâche avec l'id id
 */
export async function getTaskById(id) {
  try {
    const response = await axios.get(`${url}/${id}`, {
      headers: {
        'Authorization': `Bearer ${getAuthToken()}`
      }
    });
    return response.data; 
  } catch (error) {
    console.error('Erreur lors de la récupération de la tâche:', error.response ? error.response.data : error.message);
    throw error; 
  }
}

/**
 * Récupère la liste de tâches associés à l'email de l'utilisateur
 * 
 * @param {string} userEmail - l'email de l'utilisateur associé aux tâches à récupérer
 * @returns response.data - la liste de tâches
 */
export async function getTasksByUserEmail(userEmail) {
  try {
    const response = await axios.get(`${url}/user/${userEmail}`, {
      headers: {
        'Authorization': `Bearer ${getAuthToken()}` 
      }
    });
    return response.data;
  } catch (error) {
    console.error('Erreur lors de la récupération des tâches de l\'utilisateur:', error.response ? error.response.data : error.message);
    throw error;
  }
}

/**
 * Supprimer la tâche avec l'id id
 * 
 * @param {number} id - l'id de la tâche à supprimer
 */
export async function deleteTask(id) {
  try {
    await axios.delete(`${url}/${id}`, {
      headers: {
        'Authorization': `Bearer ${getAuthToken()}`
      }
    });
    console.log(`Tâche ${id} supprimée avec succès`);
  } catch (error) {
    console.error('Erreur de suppression de la tâche:', error.response ? error.response.data : error.message);
    throw error; 
  }
}

/**
 * Supprime la tâche avec la date de création
 * 
 * @param {Date} createDate - la date de la tâche à supprimer
 * @returns true si supprimer, false sinon.
 */
export const deleteTaskByCreateDate = async (createDate) => {
  try {
    const response = await axios.delete(`${url}/delete`, {
      params: { createDate },
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${getAuthToken()}` 
      }
    });
    return response.data;
  } catch (error) {
    console.error("Erreur lors de la suppression de la tâche:", error);
    throw error;
  }
};

/**
 * Récupère la liste des tâches d'un projet.
 * 
 * @param {number} projectId - l'id du projet associé aux tâches
 * @returns liste de tâches
 */
export const getTasksByProjectId = async (projectId) => {
  try {
      const response = await axios.get(`${url}/project/${projectId}`, {
        headers: {
          'Authorization': `Bearer ${getAuthToken()}` // Add Bearer token to the headers
        }
      });
      return response.data; // Renvoie les tâches récupérées
  } catch (error) {
      console.error("Erreur lors de la récupération des tâches par projet:", error);
      throw error;
  }
};